core/src/main/scala/org/json4s/Extraction.scala

```diff
@@ -667,6 +667,9 @@ object Extraction {
       val deserializer = formats.typeHints.deserialize
       if (!deserializer.isDefinedAt(typeHint, obj)) {
         val concreteClass = formats.typeHints.classFor(typeHint) getOrElse fail("Do not know how to deserialize '" + typeHint + "'")
+        if (!typeInfo.erasure.isAssignableFrom(concreteClass)) {
+          fail(s"Type hint $typeHint which refers to class ${concreteClass.getName} cannot be extracted as an instance of ${typeInfo.erasure.getName}")
```

Comment:
@magnolia-k @mccartney I'm wondering if it would be better to throw a `MappingException` with of the following as the `cause`
1) a `ClassCastException` with a message resembling this
2) a custom exception which holds the 2 incompatible `Class` instances
3) same as `2` but extending `ClassCastException`

Reply:
Alternately, for maximum backward-compatibility,  we could just throw a `ClassCastException`, which is what happens today, assuming the return value of `extract` is stored in a variable declared as the type incompatible with the type hint class.

Reply:
I don't see any preference between the options presented, to be honest.

Reply:
If you have no preference, I think the simple mapping exception seems okay for now. Can we get this PR merged?